### PR TITLE
feat(#264): Fix Effectiveness split-window comparison (Phase 3)

### LINKED ISSUE #264: Self-improving pipeline Phase 3: feedback loop closure via split-window comparison
## Parent Issue

Child of #259 (Self-improving pipeline). See #259 for full technical design.

## Customer Problem

After the pipeline detects a recurring problem and proposes a fix, the user has no way to know whether that fix actually worked. They merge the improvement PR on faith, but subsequent health reports don't tell them "the pattern you fixed is now appearing less often." Without this feedback, the user can't distinguish effective fixes from ineffective ones, can't prioritize which types of improvements to invest in, and can't build confidence that the pipeline is genuinely getting smarter over time.

## User Segments

- **Solo developer**: Wants to see that the time spent merging improvement PRs actually paid off — "this pattern dropped from 40% to 10% after my fix."
- **Team lead**: Needs evidence that the self-improving pipeline is delivering real ROI — which fixes worked, which didn't, and where to focus next.

### Current Journey (Phase 1+2 shipped)

1. User merges an improvement PR that targets a recurring pattern
2. User continues running the pipeline on new issues
3. Health reports show overall sustain rates and pattern alerts, but nothing connects these trends to the specific fix
4. User mentally compares health reports across time — imprecise, tedious, easy to miss changes
5. User has no evidence to decide whether the fix worked or whether to try a different approach

### Target Journey (Phase 3 shipped)

1. User merges an improvement PR that targets a recurring pattern
2. Pipeline automatically discovers the fix merge date and links it to the targeted pattern
3. While fewer than 5 post-fix PRs exist, health report shows "insufficient data — N of 5 required PRs analyzed" with clear progress
4. Once 5+ post-fix PRs exist, health report shows "Fix Effectiveness: pattern X sustain rate 38% → 12% (improved, 8 PRs analyzed)" with direct attribution to the fix issue
5. User can see at a glance which fixes worked, which didn't, and where to re-invest — no manual tracking or guesswork

### S1 — Fix effectiveness appears in health report after sufficient data (Functional)

Given an improvement issue was merged to address a recurring pattern, and at least 5 PRs with telemetry have been merged since the fix
When the user runs the health report
Then the "Fix Effectiveness" section shows the pattern name, the fix issue link, the before/after sustain rate, and a directional indicator (improved / unchanged / worsened)

### S2 — Insufficient post-fix data is reported honestly (Functional)

Given an improvement issue was merged but fewer than 5 PRs with telemetry have been merged since the fix
When the user runs the health report
Then the "Fix Effectiveness" section shows "insufficient data" for that pattern rather than a misleading delta, along with how many post-fix PRs have been analyzed so far

### S3 — Fix merge date is discovered and cached automatically (Functional)

Given a proposals_emitted entry has a fix_issue_number but no cached fix_merged_at date
When the aggregation script runs
Then it discovers the merge date via GitHub API query, caches it in the calibration data for future runs, and does not re-query on subsequent runs

### S4 — User can see which fixes worked and which didn't (Intent)

Given the pipeline has proposed and merged multiple improvement fixes over time
When the user reviews the health report's Fix Effectiveness section
Then they can distinguish effective fixes (pattern frequency dropped) from ineffective ones (pattern frequency unchanged or worsened), enabling informed decisions about what types of improvements to invest in — without manual tracking or guesswork

### From Experience-Owner Framing (D-264-1 through D-264-4)

- **D-264-1 — Fix attribution through existing proposals_emitted linkage**: Reuse the `fix_issue_number` field written at proposal creation (Phase 2) rather than introducing new tracking
- **D-264-2 — Equal decay weighting for fair comparison**: Both before-fix and after-fix windows use the same global decay relative to now (per D-259-12)
- **D-264-3 — Minimum data threshold before reporting**: Require ≥5 effective PRs in the post-fix window before showing a delta
- **D-264-4 — Directional language over numbers alone**: Report "improved / unchanged / worsened / insufficient data" alongside rates

### From Solution-Designer Technical Design (D-264-5 through D-264-11)

- **D-264-5 — Post-accumulation per-pattern split with enriched contributions**: Build fix effectiveness after the PR accumulation loop completes, using enriched `$prContributions` that carry per-category data. For each `proposals_emitted` entry with `fix_issue_number`, discover/cache `fix_merged_at`, split `$prContributions` at the merge date, filter to the targeted category, and compute before/after sustain rate.
- **D-264-6 — Per-fix `gh pr list --search` with cache-on-first-discovery**: Query `gh pr list --repo {repo} --state merged --search "closes #{N} OR fixes #{N} OR resolves #{N}" --json number,mergedAt --sort updated --limit 5` per uncached entry. Pick the result with the latest `mergedAt`. Cache the result; no re-query after caching.
- **D-264-7 — Honest status for unmerged fixes**: If `gh pr list` returns empty, don't set `fix_merged_at`. Re-query on next run. When proposals_emitted has entries with `fix_issue_number` but none have `fix_merged_at`, the health report shows a placeholder: "Fix Effectiveness: awaiting fix merge (N proposals pending)" rather than silently omitting the section.
- **D-264-8 — Table format for Fix Effectiveness section**: Markdown table consistent with existing health report sections.
- **D-264-9 — 5% deadzone with plain directional labels**: Same threshold as overall trend indicator (D-259-17). `improved` = after < before − 0.05; `worsened` = after > before + 0.05; `unchanged` = within ±5%; `insufficient data` = fewer than 5 post-fix PRs.
- **D-264-10 — Enrich $prContributions with per-category hashtable**: Add a `categories` hashtable to each `$prContributions` entry with per-category `wTotal`/`wAccepted`. **Scope boundary**: this enrichment powers Fix Effectiveness only — the Category Hotspots per-category trend column remains deferred (existing `—` placeholder). A future PR may wire this data into Hotspots as a separate scope decision.
- **D-264-11 — Normal-run-only write-back for fix_merged_at**: `fix_merged_at` is cached during normal (non-HealthReport) runs alongside existing `proposals_emitted` write-back, consistent with D-259-15 (HealthReport is read-only). A new `$fixMergedAtChanged` flag is added to the write-back gate condition to ensure mutations to existing entries trigger persistence.

## Customer Surface Identification

- **Surface type**: CLI terminal — health report Markdown output from `Invoke-AggregateReviewScores -HealthReport`, displayed by Process-Review or read directly by the user
- **Secondary surface**: GitHub issue bodies (fix issue links in the report)

## Design Intent

Enable users to see measurable evidence that pipeline improvement fixes actually work — closing the feedback loop from "detect pattern → propose fix → merge fix → verify improvement" — so they can build justified confidence in the self-improving pipeline and make informed decisions about future investments.

## CE Gate Readiness

1. **Surface type**: CLI (PowerShell script terminal output)
2. **Tool availability**: Terminal for script execution, `gh` CLI for GitHub queries
3. **Manual fallback**: Run `Invoke-AggregateReviewScores -HealthReport` in terminal and inspect the "Fix Effectiveness" section manually; verify calibration JSON for `fix_merged_at` caching
4. **Draft scenarios**: S1–S4 above (3 functional, 1 intent)

---

### Refined Design Goal

Add a "Fix Effectiveness" section to the existing health report that measures whether merged improvement fixes actually reduced their targeted pattern's frequency, using decay-weighted split-window comparison — extending the existing aggregation infrastructure with per-category contribution tracking and merge-date discovery.

### Existing Infrastructure (Already Built)

| Capability | Where | Status |
|---|---|---|
| Health report with 5 sections (sustain rate, hotspots, depth, D10, systemic) | `Format-HealthReport` in `aggregate-review-scores-core.ps1` | ✅ |
| Overall temporal split with median-index partitioning | Lines 607-623 of `aggregate-review-scores-core.ps1` | ✅ |
| Per-PR contribution capture (`wTotal`, `wAccepted`, `mergedAt`) | `$captureContribution` scriptblock, lines 485-497 | ✅ |
| `proposals_emitted` read/write with unknown-field preservation (D-263-7) | Lines 328-329 (read), 1043 (write-back) | ✅ |
| `fix_issue_number` written at proposal creation | `Update-CIICalibrationLinkage` in `create-improvement-issue-core.ps1` | ✅ |
| `$GhCliPath` parameter for test-time mock injection | Line 205 | ✅ |
| Exponential decay weighting (λ=0.023, half-life ~30 days) | Lines 514-520 | ✅ |
| HealthReport read-only mode (suppresses all write-back) | Line 1023 gate condition | ✅ |

#### 1. Per-Category Contribution Enrichment (D-264-10)

Enrich the existing `$captureContribution` scriptblock to snapshot per-category effective counts before each PR and compute per-category deltas after:

```
$prContributions entry (enriched):
{
    mergedAt:   "2026-03-15T10:00:00Z"
    wTotal:     2.3        # aggregate (unchanged)
    wAccepted:  1.8        # aggregate (unchanged)
    categories: {
        "security":    { wTotal: 0.8, wAccepted: 0.3 }
        "architecture": { wTotal: 0.5, wAccepted: 0.5 }
        ...
    }
}
```

**Implementation note**: snapshot per-category `effectiveCount` and `sustained` as scalar pairs before each PR loop iteration (not deep-copy of the mutable `$ctx.categoryData` hashtable). Reset snapshot at loop start. The `$captureContribution` scriptblock is dot-sourced and shares loop-local variables — the per-category snapshot variables must be explicitly initialized before each PR to avoid stale data from the previous PR.

#### 2. Merge-Date Discovery and Caching (D-264-6, D-264-7, D-264-11)

**Discovery flow** (post-accumulation, before health report generation):

```
For each proposals_emitted entry:
  if has fix_issue_number AND no fix_merged_at:
    result = gh pr list --repo {repo} --state merged \
      --search "closes #{N} OR fixes #{N} OR resolves #{N}" \
      --json number,mergedAt --sort updated --limit 5
    if result non-empty:
      pick entry with latest mergedAt
      set fix_merged_at = mergedAt (entry mutation via Add-Member)
      set $fixMergedAtChanged = $true
    else:
      skip (re-query on next run)
```

**Caching**: `fix_merged_at` is an ISO-8601 string added to the `proposals_emitted` entry via `Add-Member` (same pattern as `fix_issue_number` in `create-improvement-issue-core.ps1`). Persisted during normal-run write-back only.

**Write-back gate**: The existing write-back condition `($depthStateChanged -or $proposalsChanged -or $complexityHistoryChanged)` is extended to include `$fixMergedAtChanged`.

#### 3. Split-Window Fix Effectiveness Computation (D-264-5, D-264-9)

**New private function**: `Compute-FixEffectiveness`

```
Input:
  - $ProposalsEmitted: array of proposal entries (with fix_merged_at)
  - $PrContributions: enriched per-PR contributions
  - $Deadzone: 0.05 (5%)
  - $MinPostFixPrs: 5

Output: array of results:
  {
    pattern_key: "instruction:security"
    fix_issue_number: 263
    before_rate: 0.38
    after_rate: 0.12
    indicator: "improved"     # improved | worsened | unchanged | insufficient data | no before data | awaiting merge
    post_fix_prs: 8
  }
```

**Algorithm per proposals_emitted entry with fix_merged_at**:

1. Extract category from `pattern_key` (format: `{fix_type}:{category}`)
2. Partition `$PrContributions` at `fix_merged_at`:
   - Before-window: entries with `mergedAt < fix_merged_at`
   - After-window: entries with `mergedAt >= fix_merged_at`
3. **Stacked-fix windowing** (D-264-6 amendment): If multiple fixes share the same `pattern_key`, cap each fix's after-window at the next fix's `fix_merged_at`. Latest fix's after-window extends to now.
4. Count post-fix PRs. If < 5: indicator = "insufficient data", skip rate computation
5. Accumulate per-category `wTotal`/`wAccepted` for both windows from the `categories` hashtable
6. **Before-window empty** (category wTotal = 0): indicator = "no before data"
7. **After-window zero findings** (category wTotal = 0): effective sustain rate = 0% (pattern absent = best possible outcome)
8. Compute sustain rates: `before_rate = wAccepted / wTotal`, `after_rate = wAccepted / wTotal`
9. Apply deadzone: if after < before − 0.05 → "improved"; if after > before + 0.05 → "worsened"; else → "unchanged"

#### 4. Fix Effectiveness Report Section (D-264-8)

Added to `Format-HealthReport` after Systemic Pattern Alerts.

**Full data available**:
```markdown
## Fix Effectiveness

| Pattern | Fix | Before | After | Δ | PRs |
|---------|-----|--------|-------|---|-----|
| instruction:security | #263 | 38% | 12% | improved | 8 |
| agent-prompt:implementation-clarity | #265 | 22% | 20% | unchanged | 6 |
| skill:pattern | #270 | — | — | insufficient data (2/5) | 2 |
```

**Awaiting merge** (proposals exist but no merged fix found):
```markdown
## Fix Effectiveness

Awaiting fix merge (2 proposals pending).
```

**No before data**:
```markdown
| instruction:security | #263 | — | 12% | no before data | 8 |
```

**Section omitted entirely**: Only when no proposals_emitted entries have `fix_issue_number` at all.

#### 5. Health Report Context Extension

New key in `$healthReportContext`:

```powershell
$healthReportContext = @{
    # ... existing 14 keys ...
    FixEffectiveness = $fixEffectivenessResults  # array from Compute-FixEffectiveness
}
```

### proposals_emitted Schema (Post Phase 3)

```json
{
  "pattern_key": "instruction:security",
  "evidence_prs": [245, 248, 252],
  "first_emitted_at": "2026-04-03T10:00:00Z",
  "fix_issue_number": 263,
  "fix_merged_at": "2026-04-10T14:30:00Z"
}
```

Fields added by each phase:
- Phase 1: `pattern_key`, `evidence_prs`, `first_emitted_at` (written by aggregate-review-scores-core.ps1)
- Phase 2: `fix_issue_number` (written by create-improvement-issue-core.ps1)
- Phase 3: `fix_merged_at` (written by aggregate-review-scores-core.ps1, cached after first gh query)

### System Touchpoints

| File | Change |
|---|---|
| `aggregate-review-scores-core.ps1` | Enrich `$captureContribution` with per-category data; add `Compute-FixEffectiveness` private function; add merge-date discovery loop; add `$fixMergedAtChanged` to write-back gate; pass `FixEffectiveness` to health report context |
| `Format-HealthReport` (in same file) | Add "Fix Effectiveness" section after Systemic Pattern Alerts |
| `Tests/aggregate-review-scores.Tests.ps1` | Split-window math tests; insufficient data threshold; merge date caching and re-query skip; before/after sustain rate; stacked-fix windowing; zero-findings-post-fix; awaiting-merge placeholder |

### Edge Cases (Resolved)

1. **No proposals_emitted with fix_issue_number**: Fix Effectiveness section omitted entirely
2. **Fix not yet merged** (D-264-7): Re-query next run; show "awaiting fix merge (N proposals pending)" placeholder
3. **< 5 post-fix PRs**: Show "insufficient data (N/5)" with current count
4. **Pattern no longer appears post-fix** (zero findings in category): Effective sustain rate = 0%, report as "improved" — pattern absence is the best outcome
5. **Multiple fixes for same pattern_key**: Each fix evaluated in its own bounded window (`[fix_N.merged_at, fix_N+1.merged_at)` for intermediate, `[fix_last.merged_at, now)` for latest)
6. **Before-window empty** (no findings for category in pre-fix window): Report "no before data" instead of a comparison
7. **Write-back preserves all fields**: `fix_merged_at` added via entry mutation (Add-Member), same pattern as `fix_issue_number` in D-263-7

### Adversarial Design Challenge Summary

Three challenge passes (2 feasibility + 1 product-alignment) produced 15 findings after deduplication. All addressed:

1. **`gh` search keyword gap (high)**: Resolved by D-264-6 amendment — search for `closes OR fixes OR resolves` with `--sort updated`
2. **Write-back gate (high)**: Resolved — `$fixMergedAtChanged` flag added to gate condition
3. **Edge case 4/6 contradiction (high)**: Resolved — zero post-fix findings = 0% sustain rate = "improved" (pattern absence is best outcome)
4. **Per-category snapshot complexity (medium)**: Noted — scalar snapshots, explicit reset per PR iteration
5. **Sustain rate vs count metric (medium)**: Dismissed — rate is the established pipeline metric; consistent with user expectations
6. **Stacked-fix overlapping windows (medium)**: Resolved — bounded windows per fix
7. **Before-window empty (medium)**: Resolved — "no before data" label
8. **Silent omission inconsistency (medium)**: Resolved by D-264-7 amendment — "awaiting fix merge" placeholder
9. **D-264-10 scope vs deferred per-category trend (medium)**: Resolved — explicit scope boundary statement in D-264-10
10. **`--limit 1` ordering (medium)**: Resolved — `--limit 5` + pick latest mergedAt
11. **Per-category insufficient-data qualifier (medium)**: Dismissed — 5-PR minimum is pragmatic; per-category qualification adds complexity without proportional benefit
12. **Write-back entry mutation vs array replacement (medium)**: Clarified — entry mutation, same as D-263-7
13. **HealthReport repeated API cost (low)**: Dismissed — negligible for 0-5 proposals
14. **Growing context hashtable (low)**: Dismissed — stable extension pattern
15. **Phase 2 deferral comment staleness (medium)**: Noted for implementation — update comment in Format-HealthReport and code-review.md

No escalations remain.

### Rejected Alternatives

- **Batch `gh` search for multiple fixes**: Rejected — unreliable OR semantics across `closes` references; premature optimization for 0-5 uncached entries
- **Track failed lookup attempts**: Rejected — over-engineering; re-query cost is negligible for small proposal counts
- **Narrative format for Fix Effectiveness**: Rejected — doesn't scale beyond 2-3 patterns; inconsistent with the rest of the report's table-based structure
- **Confidence-weighted deadzone threshold**: Rejected — harder to explain to users; the ≥5 PR minimum provides sufficient floor
- **Separate per-category contribution list**: Rejected — parallel lists to keep in sync; no benefit over enriching existing structure
- **Allow HealthReport to write back fix_merged_at**: Rejected — violates D-259-15 (HealthReport is read-only); adds special cases to write-back suppression
- **Re-scan raw findings instead of enriching prContributions**: Rejected — requires reconstructing temporal partitioning from data that doesn't carry per-PR timestamps

---

## Scope

Phase 3 of 3. Implements feedback loop closure — measures whether merged improvement fixes actually reduced the targeted pattern's frequency using decay-weighted split-window comparison.

## Dependencies
- Phase 1 (#259 health report / `Format-HealthReport` function) must ship first ✅ merged
- Phase 2 (#263 automated proposals) must ship first — provides `fix_issue_number` in `proposals_emitted` ✅ merged
- Requires enough post-fix merged PRs to have statistical power

## Acceptance Criteria

- [ ] Split-window comparison reports before/after sustain rate for patterns with `fix_issue_number` and >=5 post-fix effective PRs
- [ ] `fix_merged_at` is cached in `proposals_emitted` after first `gh` query
- [ ] Health report includes "Fix Effectiveness" section with merged-fix attribution
- [ ] Insufficient post-fix data explicitly reported rather than showing misleading deltas
- [ ] Global decay relative to now applied to both window halves
- [ ] `gh` search covers `closes`, `fixes`, and `resolves` keywords (D-264-6)
- [ ] Stacked fixes for same pattern_key evaluated in bounded windows (D-264-6 amendment)
- [ ] "Awaiting fix merge" placeholder shown when proposals exist but no merged fix found
- [ ] Per-category contribution enrichment scoped to Fix Effectiveness only (D-264-10 boundary)
- [ ] Write-back gate triggers for `fix_merged_at` mutations via `$fixMergedAtChanged` flag

## Testing Scope
- Unit (Pester): split-window math; insufficient data threshold; merge date caching and re-query skip; before/after sustain rate calculation; stacked-fix windowing; zero-findings-post-fix edge case; awaiting-merge placeholder; no-before-data label; per-category contribution capture correctness

### PR DESCRIPTION
**Pull request by Grimblaz**
﻿## Summary

Implements Phase 3 of the self-improving review pipeline (issue #264): **Fix Effectiveness split-window comparison**. This adds the ability to measure whether systemic pattern fixes actually improved review outcomes by comparing category-level finding rates in a before-fix window vs an after-fix window.

### What it does

When a systemic pattern fix has been merged, the health report now shows whether each affected category improved, worsened, or stayed unchanged ΓÇö with a 5% deadzone to filter noise. The system discovers merge dates from GitHub, caches them in calibration data, and computes per-category contribution deltas across the fix boundary.

### Key design decisions (from issue #264)

- **D-264-5**: Per-PR category contribution hashtable enrichment (stacked per-proposal)
- **D-264-6**: Dedicated `Measure-FixEffectiveness` function with `$Deadzone` and `$MinPostFixPrs` parameters
- **D-264-7**: Zero-cost path ΓÇö no merge-date lookup when no `fix_issue_number` exists
- **D-264-8**: Lazy discovery loop with `fix_merged_at` caching + write-back gate
- **D-264-9**: `healthReportContext` dictionary carries fix effectiveness to Format-HealthReport
- **D-264-10**: D-264-10 documentation references in code-review.md
- **D-264-11**: Deep-copy snapshot of category data at the fix boundary

Closes #264

## Changed Files

| File | Change |
| --- | --- |
| `.github/scripts/lib/aggregate-review-scores-core.ps1` | Core implementation: `Measure-WindowCategoryTotals`, `Measure-FixEffectiveness`, per-category contribution enrichment, merge-date discovery loop, write-back gate, Format-HealthReport fix effectiveness section |
| `.github/scripts/Tests/aggregate-review-scores.Tests.ps1` | 34 new tests across 5 Describe blocks: contribution enrichment (4), Measure-FixEffectiveness (10), merge-date discovery (7), Format-HealthReport rendering (8), integration (3), plus requires-gh live test (MF4) |
| `Documents/Design/code-review.md` | Phase 3 documentation: Fix Effectiveness row in health report table, new subsection with helper function docs and discovery loop description |

### Tier 1 ΓÇö Build & Validate
- **PSScriptAnalyzer**: 0 violations
- **markdownlint-cli2**: 0 errors (summary output on commit)
- **Structural checks**: all passing (0 legacy references)
- **Pester (no-gh)**: 66/66 tests pass (targeted selection)
- **Pester (full suite)**: 328 passed, 24 failed (all pre-existing on main), 1 skipped
- **Pre-existing failures**: ~24 tests in systemic_patterns and prosecution_depth blocks fail identically on both main and feature branch ΓÇö confirmed NOT regressions

### Tier 2 ΓÇö Structural
- `measure-guidance-complexity.ps1`: 0 agents over ceiling
- No broken references detected

### Tier 3 ΓÇö Strength
- 34 new tests covering: happy path, edge cases (insufficient data, no-before-data, zero findings), stacked fixes, deadzone boundaries, merge-date caching, write-back round-trip, Format-HealthReport rendering variants

## Adversarial Review Scores

| Stage | Prosecutor | Defense | Judge rulings |
| --- | --- | --- | --- |
| Code Review | 19 pts (6 sustained) | 1 pt (1 disproved, 0 rejected) | 7 rulings |
| Post-fix Review | 0 pts (0 sustained) | ΓÅ¡∩╕Å N/A | ΓÅ¡∩╕Å N/A |
| CE Review | 2 pts (2 sustained) | 1 pt (1 disproved, 0 rejected) | 3 rulings |

### Code Review Findings

| ID | Severity | Category | Ruling | Fix |
| --- | --- | --- | --- | --- |
| MF1 | Critical | script-automation | Γ£à Sustained | `--json number, mergedAt` comma-operator bug ΓåÆ quoted `--json 'number,mergedAt'` |
| MF2 | Low | implementation-clarity | Γ¥î Defense sustained | Empty-string `fix_merged_at` path ΓÇö no code path produces this |
| MF3 | Low | implementation-clarity | Γ£à Sustained | Hardcoded `/5` ΓåÆ thread `min_post_fix_prs` through result |
| MF4 | Medium | script-automation | Γ£à Sustained | Missing requires-gh live test ΓåÆ added with PESTER_LIVE_GH opt-in |
| MF5 | Low | script-automation | Γ£à Sustained | Missing `$LASTEXITCODE` check after gh call ΓåÆ added with Write-Warning |
| MF6 | Low | implementation-clarity | Γ£à Sustained | Redundant `$dAccepted -gt 0` guard ΓåÆ simplified to `$dTotal -gt 0` |
| MF7 | Low | pattern | ≡ƒôï DEFERRED-SIGNIFICANT | Private helper naming convention (ARS prefix) ΓåÆ #285 |

### Post-fix Review
Triggered (MF1 Critical severity). 1 prosecution pass, 0 findings ΓÇö clean.

### CE Gate

**Result**: Γ£à CE Gate passed ΓÇö intent match: strong

**Scenarios exercised**:
- S1: Fix effectiveness data appears in health report with correct before/after/delta values ΓÇö Γ£à Passed
- S2: Insufficient post-fix data shows clear indicator with PR threshold ΓÇö Γ£à Passed
- S3: Merge-date discovery fires on first run and caches on subsequent runs ΓÇö Γ£à Passed
- S4: Improved vs worsened vs unchanged categories are visually distinguishable ΓÇö Γ£à Passed

**CE prosecution**: 3 findings (F1, F2: documentation nits sustained; F3: untested error path disproved)

**Process-Review**: no systemic gap found

## Prosecution Depth Summary

| Category | Depth | Rationale |
| --- | --- | --- |
| architecture | full | ΓÇö |
| security | full | ΓÇö |
| performance | full | ΓÇö |
| pattern | full | ΓÇö |
| implementation-clarity | full | ΓÇö |
| script-automation | full | ΓÇö |
| documentation-audit | full | ΓÇö |

Re-activated categories: none

## Process Gaps

- **MF7 ΓåÆ #285**: Private helper naming convention ΓÇö all Phase 3 helpers lack ARS prefix (tech-debt cleanup)
- **No other systemic gaps identified** by Process-Review during CE Gate

### Slowdowns
1. **VS Code lockup** mid-session required full restart and smart resume from session memory
2. **MF1 (Critical)**: `--json number, mergedAt` ΓÇö PowerShell's comma operator silently creates an array argument. Mock tests couldn't catch this because they don't validate argument format against real `gh` CLI. Discovered only by adversarial prosecution.

### Late-Failing Checks
- None. All regressions caught at Tier 1. The 24 pre-existing test failures are on main and predate this PR.

### Workflow Guardrail Improvement
- **Requires-gh live argument format tests** (MF4 pattern): Any new `gh` CLI integration should include a `requires-gh` live test that validates argument syntax against the real CLI, since mock-based tests cannot detect PowerShell argument-expansion bugs like the comma-operator issue in MF1. This pattern is now established in the test file and should be replicated for future gh integrations.

## Summary by Sourcery

Add per-category Fix Effectiveness measurement to the review aggregation pipeline, including split-window comparisons around systemic pattern fixes and surfaced results in the health report.

New Features:
- Introduce Fix Effectiveness computation that compares per-category sustain rates before and after systemic pattern fixes using proposal metadata and PR contributions.
- Render a Fix Effectiveness section in the health report showing per-fix before/after rates, deltas, indicators, and pending-fix status.

Enhancements:
- Enrich per-PR contribution data with per-category weighted totals to enable category-level fix impact analysis.
- Add a merge-date discovery loop that resolves fix issue numbers to merged PR dates via the GitHub CLI and caches them in calibration data.
- Extend the aggregation context to carry Fix Effectiveness results through to report formatting and calibration write-back logic.

Documentation:
- Document the Fix Effectiveness feature, helper functions, per-category enrichment, and merge-date discovery loop in the code review design spec.

Tests:
- Add extensive unit, rendering, and integration tests for Fix Effectiveness computation, per-category enrichment, merge-date discovery, health report formatting, and calibration round-trips, plus an opt-in live GitHub CLI argument-format test.

### CODE REVIEW

File: .github/scripts/Tests/aggregate-review-scores.Tests.ps1 [hunk 1]
```diff
@@ -3395,167 +3398,2102 @@ exit 0
...
+            -Because 'entry without fix_issue_number does not count as awaiting merge'
+    }
+}
+
+# ==================================================================
+# Describe: Fix Effectiveness — merge-date discovery loop (Step 3 #264)
+# RED tests — all fail until the merge-date discovery loop is
+# implemented inside Invoke-AggregateReviewScores. The loop iterates
+# proposals_emitted entries with fix_issue_number but no fix_merged_at,
+# queries gh CLI for the merge date, and caches the result.
```
[the remark this answers is not in the document]
Comment (Grimblaz) [reply .github/scripts/Tests/aggregate-review-scores.Tests.ps1]: **✅ ACCEPT** — Fixed in `fde4bd4`. Stale TDD scaffolding comment removed — merge-date discovery loop implementation is complete.

File: .github/scripts/Tests/aggregate-review-scores.Tests.ps1 [hunk 2]
```diff
@@ -3364,24 +3364,27 @@ findings:
 }
 
 # ==================================================================
-# Describe: Read-only mode and insufficient data handling (Step 4 #259)
+# Describe: Fix Effectiveness — per-category contribution enrichment (Step 1 #264)
+# RED tests — Tests 3, 4, 5 fail until Measure-FixEffectiveness and
+# the Fix Effectiveness health report section are implemented (Steps 2+4).
+# Tests 1 and 2 are invariant guards that should pass immediately.
```
[the remark this answers is not in the document]
Comment (Grimblaz) [reply .github/scripts/Tests/aggregate-review-scores.Tests.ps1]: **✅ ACCEPT** — Fixed in `fde4bd4`. Stale TDD scaffolding comment removed — implementation is complete.

File: .github/scripts/lib/aggregate-review-scores-core.ps1 [hunk 3]
```diff
@@ -518,6 +760,14 @@ function Invoke-AggregateReviewScores {
         $prWtotalBefore = $ctx.weightedTotal
         $prWacceptedBefore = $ctx.weightedAccepted
 
+        # Snapshot per-category data for per-category enrichment (Phase 3 D-264-1)
+        # Deep-copy scalar values — inner hashtables are mutable references (MF-1).
+        $prCatBefore = @{}
+        foreach ($catKey in @($ctx.categoryData.Keys)) {
+            $catData = $ctx.categoryData[$catKey]
+            $prCatBefore[$catKey] = @{ effectiveCount = $catData.effectiveCount; sustained = $catData.sustained }
+        }
```
[the remark this answers is not in the document]
Comment (Grimblaz) [reply .github/scripts/lib/aggregate-review-scores-core.ps1]: **❌ REJECT** — "(MF-1)" at L764 is a single inline parenthetical note. No conflicting "MF1" (without hyphen/parens) reference exists anywhere in the codebase — this is a phantom inconsistency. The reference style is informal shorthand, not a cross-reference to PR metadata IDs.

File: .github/scripts/lib/aggregate-review-scores-core.ps1 [hunk 4]
```diff
@@ -14,7 +14,7 @@ $script:_ARSCoreLibDir = Split-Path -Parent $PSCommandPath
 function Get-FlexProperty {
     param($Object, [string]$Name)
     if ($null -eq $Object) { return $null }
-    if ($Object -is [hashtable]) { return $Object[$Name] }
+    if ($Object -is [System.Collections.IDictionary]) { return $Object[$Name] }
     if ($Object.PSObject.Properties.Name -contains $Name) { return $Object.$Name }
```
[the remark this answers is not in the document]
Comment (Grimblaz) [reply .github/scripts/lib/aggregate-review-scores-core.ps1]: **✅ ACCEPT** — Fixed in `fde4bd4`. Comment updated to "PSCustomObject/IDictionary" to match the actual `[System.Collections.IDictionary]` type check.
